### DISCUSSION TITLE
Empty attribute with relationship gives error when accessing outside of session

### DISCUSSION THREAD

**Discussion by LukasK13** [sqlalchemy/sqlalchemy#8506] [Usage Questions] [answered]
I'm using sqlalchemy 1.4.36 with MariaDB 10.5.13.

I'm sticking to the example used in the sqlalchemy ORM tutorial for clarity.
I define up two object models with a one-to-many relationship (User and Address). The only difference to the tutorial is that I set `lazy='subquery'` for User and open the session with `expire_on_commit=False`. Afterwards I instanciate a User _spongebot_ with an address and a User _patrick_ without an address. Before writing the users to the database I'm accessing the child attribute _addresses_ which of course returns an empty list for_patrick_ and the defined address for _spongebob_.
However accessing the attribute after writing the objects to the database outside of the _session_-context, I get the error `sqlalchemy.orm.exc.DetachedInstanceError: Parent instance <User at 0x7f211864e0d0> is not bound to a Session; lazy load operation of attribute 'addresses' cannot proceed (Background on this error at: https://sqlalche.me/e/14/bhk3)` for _patrick_. _spongebob_ still returns the correct address.
(I think) I understand the concept of lazy-loading, but in this case the objects and all childs have been available before accessing the database and therefore should remain available after closing the session. This works for _spongebob_ which has an address, but does not work for _patrick_ without an address. I would expect the call `patrick.addresses` to return an empty list even after the session has been closed.
Adding `session.refresh(patrick)` after the commit fixes the problem.
Am I doing anything wrong or is this indeed a bug?

The reproducable code is below:
```py
from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import VARCHAR
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy import create_engine
from sqlalchemy.orm import Session

Base = declarative_base()


class User(Base):
    __tablename__ = "user_account"
    id = Column(Integer, primary_key=True)
    name = Column(VARCHAR(30))
    fullname = Column(VARCHAR(100))
    addresses = relationship(
        "Address", back_populates="user", cascade="all, delete-orphan", lazy='subquery'
    )

    def __repr__(self):
        return f"User(id={self.id!r}, name={self.name!r}, fullname={self.fullname!r})"


class Address(Base):
    __tablename__ = "address"
    id = Column(Integer, primary_key=True)
    email_address = Column(VARCHAR(200), nullable=False)
    user_id = Column(Integer, ForeignKey("user_account.id"), nullable=False)
    user = relationship("User", back_populates="addresses")

    def __repr__(self):
        return f"Address(id={self.id!r}, email_address={self.email_address!r})"


engine = create_engine("mariadb+mariadbconnector:/scott:tiger@127.0.0.1:3306/H2Monitoring",
                       echo=True, future=True)
# Base.metadata.create_all(engine)


spongebob = User(
        name="spongebob",
        fullname="Spongebob Squarepants",
        addresses=[Address(email_address="spongebob@sqlalchemy.org")],
    )
patrick = User(name="patrick", fullname="Patrick Star")
with Session(engine, expire_on_commit=False) as session:
    session.add_all([spongebob, patrick])
    print(spongebob.addresses)  # Works and gives [Address(id=1, email_address='spongebob@sqlalchemy.org')]
    print(patrick.addresses)  # Works and gives an empty list []
    session.commit()
print(spongebob.addresses)  # Works and gives [Address(id=1, email_address='spongebob@sqlalchemy.org')]
print(patrick.addresses)   # Does not work and gives error Parent instance <User at 0x7f211864e0d0> is not bound to a Session;
```

**Comment by CaselIT** [comment]
Hi,

I've edited it, but it looked like you posted a real url string. I would suggest changing the credentials :)

**Reply by LukasK13** [reply]
Wow that was incredibly stupid. Thanks a lot for the edit and the heads up!

**Comment by zzzeek** [comment, marked as answer]
the `patrick.addresses` list in the example is not assigned at all.   SQLAlchemy's ORM behavior is that ORM-mapped attributes, when unassigned on a [pending](https://docs.sqlalchemy.org/en/14/glossary.html#term-pending) object, a default value is returned, but not actually present on the object.  Normal Python behavior would be that `AttributeError` is thrown, but the ORM here wants to create some semblance of "row" -like behavior.   For collections, it returns an empty collection, but this isn't actually part of the object's state (in much older versions of SQLAlchemy, it was, but this was changed as it turned out to be a bad idea to change the state of a pending object when a getter is called).

so after you print `patrick.addresses`, there's still no `addresses` collection on `patrick`.     so after the commit, `patrick` moves to [persistent](https://docs.sqlalchemy.org/en/14/glossary.html#term-persistent), and now the "return an empty collection on access" behavior is gone; patrick has a database identity, which means to print out `.addresses` we need to load from the DB, there's no session, and it fails.

Here's other ways to change the behavior of the test:  1. assign addresses=[] to patrick up front.   2. session.flush() the objects before you first access patrick.addresses

What's the bigger picture?  bigger picture is avoid working with detached objects if possible, or if working with them, it would be in a limited scope where all the objects were fully loaded from the database, and eager loading is used up front to make sure all the collections that are to be accessed have been loaded.

**Reply by zzzeek** [reply]
I actually managed to get this behavioral element noted in the tutorial, but certainly I wouldn't expect someone to detect when this happens without doing some debugging as it's hard to spot in a real program: https://docs.sqlalchemy.org/en/14/tutorial/orm_related_objects.html#persisting-and-loading-relationships

**Reply by LukasK13** [reply]
Thank you very much @zzzeek for your detailed answer and the explanation. That fixes my problem.

**Comment by nightblure** [comment]
@zzzeek Hi!
But what if I specifically want to load some objects from the database and use them outside of the session? Is there really no way to use SQLA objects outside of a session and use relationships? Think of it as something similar to using simple dataclasses

Is there an option to detach objects from the session so that they do not try to load anything from the database when accessing the relationship attribute? in other words, we would like to rely only on what we have now downloaded from the database

**Reply by zzzeek** [reply]
> @zzzeek Hi! But what if I specifically want to load some objects from the database and use them outside of the session? Is there really no way to use SQLA objects outside of a session and use relationships? Think of it as something similar to using simple dataclasses

sure there is, like I said, "it would be in a limited scope where all the objects were fully loaded from the database, and eager loading is used up front to make sure all the collections that are to be accessed have been loaded."

if you are looking for relationships to lazy load outside of a session there is a special helper,  see [enable_relationship_loading](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.enable_relationship_loading)


> Is there an option to detach objects from the session so that they do not try to load anything from the database when accessing the relationship attribute? in other words, we would like to rely only on what we have now downloaded from the database

that's the normal behavior of a detached object, just close the session and the objects are detached

**Reply by nightblure** [reply]
hmm, but I got errors like object not bound to a session after I closed the session and accessed the relationship attributes, and I didn't want to commit, I just wanted to read previously loaded data... I couldn't find anything that would help me use previously loaded objects

**Reply by zzzeek** [reply]
you need to use [eager loading](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html)

**Reply by nightblure** [reply]
Thanks!
And will make_transient work? will it have the same effect?

**Reply by zzzeek** [reply]
make_transient doesn't detach an object, make_transient is usually used when you want to make copies of an object with a new primary key identity.

if you want to detach an object you say `session.expunge(obj)` , or `session.expunge_all()` or `session.close()` to detach everything.

that's all separate from having your data loaded up front which is the eager loading thing
